File: generated_code/GeminiPro/Java/049_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree for an integer set.
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again. It performs
 * basic operations such as insertion, search, and removal in O(log n) amortized time.
 */
public class SplayTree {

    private Node root;

    /**
     * Inner class to represent a node in the tree.
     */
    private class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key (or the last
     * accessed node if the key is not found) to the root of the tree.
     * This implementation uses the top-down splaying approach.
     *
     * @param root The root of the tree (or subtree) to splay.
     * @param key  The key to splay around.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node root, int key) {
        if (root == null || root.key == key) {
            return root;
        }

        // Create a dummy node. The right child of dummy will be the new
        // root of the left subtree, and the left child will be the new
        // root of the right subtree.
        Node dummy = new Node(0);
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;

        while (true) {
            if (key < root.key) {
                if (root.left == null) break;
                // Zig-Zig (Left-Left)
                if (key < root.left.key) {
                    root = rightRotate(root);
                    if (root.left == null) break;
                }
                // Link Right: Add current root to the right tree
                rightTreeMin.left = root;
                rightTreeMin = root;
                root = root.left;
            } else if (key > root.key) {
                if (root.right == null) break;
                // Zig-Zig (Right-Right)
                if (key > root.right.key) {
                    root = leftRotate(root);
                    if (root.right == null) break;
                }
                // Link Left: Add current root to the left tree
                leftTreeMax.right = root;
                leftTreeMax = root;
                root = root.right;
            } else { // key == root.key
                break;
            }
        }

        // Assemble the final tree
        leftTreeMax.right = root.left;
        rightTreeMin.left = root.right;
        root.left = dummy.right;
        root.right = dummy.left;

        return root;
    }

    /**
     * Searches for a key in the splay tree.
     * This operation performs a splay on the accessed node (or the last
     * node accessed during the search), moving it to the root.
     *
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree remains unchanged. After insertion,
     * the new node becomes the root of the tree.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. The closest node will become the root.
        root = splay(root, key);

        // If key is already present, do nothing
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not in the tree, the tree remains unchanged.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to the root
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring the maximum element (predecessor) to its root.
            // Splaying for 'key' in the left subtree will bring the max element to the top
            // because 'key' is greater than all elements in the left subtree.
            leftSubtree = splay(leftSubtree, key);

            // The new root of the left subtree has no right child.
            // Attach the right subtree of the original root to it.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/049_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Splay Tree class that implements a
    dictionary-like set for integers.

    The tree automatically rebalances itself by moving frequently accessed
    elements closer to the root through a process called "splaying".

    Methods:
    - insert(key): Adds an integer to the set.
    - delete(key): Removes an integer from the set.
    - search(key): Checks for an integer's presence and splays the accessed node.
    """

    class _Node:
        """A private inner class representing a node in the splay tree."""
        def __init__(self, key, parent=None, left=None, right=None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _rotate_right(self, x: _Node):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _rotate_left(self, x: _Node):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _splay(self, node: _Node):
        """
        Splays the given node to the root of the tree through a series of
        rotations.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent

            if not grandparent:
                # Zig case: Parent is the root.
                if node == parent.left:
                    self._rotate_right(parent)
                else:
                    self._rotate_left(parent)
            elif node == parent.left:
                if parent == grandparent.left:
                    # Zig-Zig case (left-left)
                    self._rotate_right(grandparent)
                    self._rotate_right(parent)
                else:
                    # Zig-Zag case (right-left)
                    self._rotate_right(parent)
                    self._rotate_left(grandparent)
            else:  # node is the right child
                if parent == grandparent.right:
                    # Zig-Zig case (right-right)
                    self._rotate_left(grandparent)
                    self._rotate_left(parent)
                else:
                    # Zig-Zag case (left-right)
                    self._rotate_left(parent)
                    self._rotate_right(grandparent)

    def search(self, key: int) -> bool:
        """
        Searches for a key in the tree.

        This method performs the splaying operation. If the key is found, the
        corresponding node is splayed to the root. If the key is not found,
        the last accessed node (the would-be parent) is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key found, splay the node and return True
                self._splay(current)
                return True

        # Key not found, splay the last visited node (if any)
        if last_visited:
            self._splay(last_visited)

        return False

    def insert(self, key: int):
        """
        Inserts a key into the tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and we're done
                self._splay(current)
                return

        # Insert new node
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key: int):
        """
        Deletes a key from the tree.

        The node containing the key (or its would-be parent) is first splayed
        to the root, and then the deletion is performed.

        Args:
            key: The integer key to delete.
        """
        # First, search for the key. This will splay the node to the root if it
        # exists, or its parent if it does not.
        self.search(key)

        # If the root does not hold the key, the key was not in the tree.
        if not self.root or self.root.key != key:
            return

        # At this point, the node to delete is the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # The left subtree exists. We will make its largest element the new root.
            # 1. Disconnect the left subtree to work on it.
            left_subtree.parent = None

            # 2. Find the maximum node in the left subtree.
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right

            # 3. Splay this maximum node to the top of the left subtree.
            # To do this, we temporarily treat the left subtree as the main tree
            # so we can reuse the _splay method.
            self.root = left_subtree
            self._splay(max_in_left)

            # 4. After splaying, `self.root` is the new root of the combined tree.
            # This new root has no right child. We attach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root